pkg/gadgettracermanager/containerutils/containerutils.go

```diff
@@ -0,0 +1,116 @@
+package containerutils
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"unsafe"
+)
+
+/*
+#define _GNU_SOURCE
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdint.h>
+
+struct cgid_file_handle
+{
+  //struct file_handle handle;
+  unsigned int handle_bytes;
+  int handle_type;
+  uint64_t cgid;
+};
+
+uint64_t get_cgroupid(char *path) {
+  struct cgid_file_handle *h;
+  int mount_id;
+  int err;
+  uint64_t ret;
+
+  h = malloc(sizeof(struct cgid_file_handle));
+  if (!h)
+    return 0;
+
+  h->handle_bytes = 8;
+  err = name_to_handle_at(AT_FDCWD, path, (struct file_handle *)h, &mount_id, 0);
+  if (err != 0)
+    return 0;
+
+  if (h->handle_bytes != 8)
+    return 0;
+
+  ret = h->cgid;
+  free(h);
+
+  return ret;
+}
+*/
+import "C"
+
+func GetCgroupID(path string) (uint64, error) {
+	cpath := C.CString(path)
+	ret := uint64(C.get_cgroupid(cpath))
+	C.free(unsafe.Pointer(cpath))
+	if ret == 0 {
+		return 0, fmt.Errorf("GetCgroupID on %q failed", path)
+	}
+	return ret, nil
+}
+
+func GetCgroup2Path(pid int) (string, error) {
+	cgroupPath := ""
+	if cgroupFile, err := os.Open(filepath.Join("/proc", fmt.Sprintf("%d", pid), "cgroup")); err == nil {
+		defer cgroupFile.Close()
+		reader := bufio.NewReader(cgroupFile)
+		for {
+			line, err := reader.ReadString('\n')
+			if err != nil {
+				break
+			}
+			if strings.HasPrefix(line, "0::") {
+				cgroupPath = strings.TrimPrefix(line, "0::")
+				cgroupPath = strings.TrimSuffix(cgroupPath, "\n")
+				break
+			}
+		}
+	} else {
+		return "", fmt.Errorf("cannot parse cgroup: %v", err)
+	}
+	if cgroupPath == "" {
+		return "", fmt.Errorf("cannot find cgroup path in /proc/PID/cgroup")
+	}
+	cgroupPath = filepath.Join("/sys/fs/cgroup/unified", cgroupPath)
+	if _, err := os.Stat(cgroupPath); os.IsNotExist(err) {
+		return "", fmt.Errorf("cannot access cgroup %q: %v", cgroupPath, err)
+	}
+
+	return cgroupPath, nil
+}
+
+func PidFromContainerId(containerID string) (int, error) {
```

Comment:
Is this Docker specific?

Reply:
Yes, it is. I assume something similar can be done with cri-o with `chroot /host crictl inspect`. We could do that in a future PR.